src/main/java/com/palantir/gradle/graal/DownloadGraalTask.java

```diff
@@ -36,50 +36,45 @@
     private static final String ARTIFACT_PATTERN = "[url]/vm-[version]/graalvm-ce-[version]-[os]-[arch].tar.gz";
     private static final String FILENAME_PATTERN = "graalvm-ce-[version]-[arch].tar.gz";
 
-    @Input private Provider<String> graalVersion;
-    @Input private Provider<String> downloadBaseUrl;
+    private final Provider<String> graalVersion;
+    private final Provider<String> downloadBaseUrl;
 
-    @TaskAction
-    public final void downloadGraal() throws IOException {
-        Path cache = getCache();
-        if (!(cache.toFile().mkdirs() || cache.toFile().exists())) {
-            throw new IllegalStateException(
-                    "Unable to make cache directory, and the cache directory does not already exist: " + cache);
-        }
-
-        InputStream in = new URL(render(ARTIFACT_PATTERN)).openStream();
-        Files.copy(in, getOutput(), StandardCopyOption.REPLACE_EXISTING);
-    }
+    @Inject
+    public DownloadGraalTask(GraalExtension extension) {
+        onlyIf(task -> !getTgz().get().exists());
+        setGroup(GradleGraalPlugin.TASK_GROUP);
+        setDescription("Downloads and caches GraalVM binaries.");
 
-    @Override
-    public final String getGroup() {
-        return GradleGraalPlugin.TASK_GROUP;
+        graalVersion = extension.getGraalVersion();
+        downloadBaseUrl = extension.getDownloadBaseUrl();
     }
 
-    @Override
-    public final String getDescription() {
-        return "Downloads and caches GraalVM binaries.";
+    @TaskAction
+    public final void downloadGraal() throws IOException {
+        Path cache = getCache().get();
+        Files.createDirectories(cache);
+        try (InputStream in = new URL(render(ARTIFACT_PATTERN)).openStream()) {
+            Files.copy(in, getTgz().get().toPath(), StandardCopyOption.REPLACE_EXISTING);
+        }
```

Comment:
Worth nothing that when aborting a gradle build running this task, the downloaded file will be corrupted and gradle won't run this task again (due to the onlyIf condition).
Solution is usually to download to a temp file, then move atomically